pyfr/solvers/navstokes/elements.py

```diff
@@ -16,15 +19,69 @@ def set_backend(self, backend, nscalupts):
                        visc_corr=visc_corr,
                        c=self.cfg.items_as('constants', float))
 
+        if self.cfg.getfloat('solver-avis', 'amu0', 0.0):
+            amu = self._avis_upts
+            tplargs.update(dict(art_vis='mu'))
+        else:
+            amu = None
+            tplargs.update(dict(art_vis='none'))
+
         if 'flux' in self.antialias:
             self.kernels['tdisf'] = lambda: backend.kernel(
                 'tflux', tplargs=tplargs, dims=[self.nqpts, self.neles],
                 u=self._scal_qpts, smats=self.smat_at('qpts'),
-                f=self._vect_qpts
+                f=self._vect_qpts, amu=amu
             )
         else:
             self.kernels['tdisf'] = lambda: backend.kernel(
                 'tflux', tplargs=tplargs, dims=[self.nupts, self.neles],
                 u=self.scal_upts_inb, smats=self.smat_at('upts'),
-                f=self._vect_upts
+                f=self._vect_upts, amu=amu
             )
+
+        # Kernels for artificial viscosity
+        if amu:
+            backend.pointwise.register('pyfr.solvers.navstokes.kernels.entropy')
+            backend.pointwise.register('pyfr.solvers.navstokes.kernels.avis')
+
+            def artf_vis():
+                # Compute entropy and save to avis_upts
+                ent = backend.kernel('entropy', tplargs=tplargs, dims=[self.nupts, self.neles],
+                                     u=self.scal_upts_inb, s=self._avis_upts)
+
+                # Compute modal coefficients of entropy and save them to avis_upts_temp
+                inVdm = np.linalg.inv(self._basis.ubasis.vdm.T)
+                inVdm = self._be.const_matrix(inVdm, tags={'inVdm', 'align'})
+                mul = backend.kernel('mul', inVdm, self._avis_upts, out=self._avis_upts_temp)
+
+                # Additional constants for element-wise artificial viscosity kernel
+                ubdegs = self._basis.ubasis.degrees
+                tplargs['c'].update(self.cfg.items_as('solver-avis', float))
+                tplargs['c'].update({'pi': np.pi, 'eps': np.finfo(backend.fpdtype).resolution})
+                tplargs.update(dict(nupts=self.nupts, nfpts=self.nfpts,
+                                    order=self._basis.order, ubdegs=ubdegs))
+
+                # Column view for avis_upts/fpts matrices
+                def col_view(mat):
+                    ioshape = mat.ioshape
+                    vshape = (ioshape[0],)
+                    nelespts = ioshape[-1]
+                    rcmap = np.array([[j, i] for i in range(ioshape[2]) for j in range(ioshape[1])])
+                    matmap = np.array([mat.mid]*nelespts)
+                    stridemap = np.array([[mat.leaddim]]*nelespts)
+
+                    return backend.view(matmap, rcmap, stridemap, vshape)
+
+                self._avis_upts_cv = col_view(self._avis_upts)
+                self._avis_fpts_cv = col_view(self._avis_fpts)
```

Comment:
Probably not safe to allocate here.  You need to get these allocated after `_avis_upts` and `_avis_fpts` have been committed but before the final commit.